### ISSUE TITLE
CS8073 false possitive for linq to SQL

### ISSUE THREAD

**Issue by msmolka** [dotnet/roslyn#57852]
**Version Used**: 
.net 5
**Steps to Reproduce**:

```c#
var query =
                from product in products
                join warehouseStock in warehouseStocks on product.Id  equals warehouseStock.ProductId  into ws
                from warehouseStock in ws.DefaultIfEmpty()
                where(warehouseStock.Id == null)
               ...

```
1.  Above query is used to generate `LEFT JOIN`
2. It generates proper SQL 
3. Comparison cannot be omitted here.

**Expected Behavior**:
Or disable for linq queries checks
Or allow suppression attribute to be used to suppress this warning.

**Actual Behavior**:
Only pragma allows disable warning.

**Comment by RikkiGibson** [comment]
At a glance this seems like an issue with the data model. If the column is nullable on the SQL side then the property `Id` that represents it should probably be of type `int?`.

Are you able to provide a self-contained reproducer?

**Comment by svick** [comment]
@RikkiGibson The column is not nullable, it's the `DefaultIfEmpty()` that makes the entity itself nullable and `WHERE warehouseStock.Id IS NULL` is how you would check that in SQL. [Here's a repro for Microsoft.EntityFrameworkCore.Sqlite.](https://gist.github.com/svick/f39ef913a584d182ad57dd36deafd3c1)

@msmolka While `warehouseStock.Id == null` corresponds to how you would do this in SQL and it does work, it doesn't actually make sense if you think about the objects involved. And if you try the same query with LINQ to Objects, you'll get a `NullReferenceException`.

What you actually want is `warehouseStock == null`. This makes sense from a logical point of view, does not produce any warnings, results in exactly the same SQL (at least based on my single test) and it also works in LINQ to Objects.

Because of that, I think the warning should stay, especially since it's a point of friction that might lead you to discover a better way to write the code. But if this is a common occurrence, maybe it would be worth it if the IDE suggested the fix for you?

**Comment by msmolka** [comment]
Ok, But I think still it should possible to disable this warning using attribute. Not only pragma.

**Comment by svick** [comment]
Note that [the documentation for the `[SuppressMessage]` attribute](https://docs.microsoft.com/en-us/dotnet/api/system.diagnostics.codeanalysis.suppressmessageattribute?WT.mc_id=DOP-MVP-5002735) explicitly says that it doesn't work on compiler diagnostics like this one.

That doesn't mean it can't change, but it does mean the current behavior is by design.

**Comment by msmolka** [comment]
ok... closing
